### PR TITLE
Customers

### LINKED ISSUE #2: Allow developers to access Customer resource
Verbs to be supported

1. GET
1. POST
1. PUT

* User should be able to GET a list of customers, and GET a single customer.
* If the query string parameter of `?_include=products` is provided, then any products that the customer is selling should be included in the response.
* If the query string parameter of `?_include=payments` is provided, then any payment types that the customer has used to pay for an order should be included in the response.
* If the query string parameter of `q` is provided when querying the list of customers, then any customer that has property value that matches the pattern should be returned.

If `/customers?q=Mic` is requested, then any customer whose first name is Michelle, or Michael, or Mick should be returned. Any customer whose last name is Michaelangelo, or Micacci, Micajkov should be returned.  Every property of the customer object should be checked for a match.

### PR DESCRIPTION
**Pull request by jeremiah3643**
## Link to Ticket
Closes #2 

## Description of Proposed Changes
- Created model for Customer
- Created mvc for Customer

## Steps to Test

Outline the steps to test

```sh
git fetch --all
git checkout customers
Add filler customers data into your SQL
Test POST, PUT, DELETE, and GET in POSTMAN
```
To further test, add q=(name you want to search) to url parameter to search by names of customers.
Add  _includes=(products or payments) to show what paymentTypes or products the customer has.

## Impacted Areas in Application

List general components of the application that this PR will affect:

* CustomersController and Customer(model)

## Mentions @username

@ricky-bruner 
@KHardt 
@askingalot 
@thatmikeparrish 

## Definition of Done

1. The project must be fully documented. This includes the following:
    1. Complete README that documents the steps to install the code, how to install any dependencies, or system configuration needed.
    1. Every class must be documented with purpose, author, and methods.
    1. Every method must be documented with purpose and argument list - which itself must contain a short purpose for each argument.
1. The project must be able to run fully, and without errors, on each teammate's system.
1. Fulfills every requirement.
Every line of code has been peer reviewed.
1. For projects that require unit testing, core functionality must be identified and have at least one test for each.

### CODE REVIEW

**Review by thatmikeparrish — APPROVED** [review]
Works for me

File: BangazonAPI/Controllers/CustomersController.cs [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+                }
+                else
+                {
+                    throw;
+                }
+            }
+        }
+        // DELETE api/customers/5
+        [HttpDelete("{id}")]
+        public async Task<IActionResult> Delete(int id)
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomersController.cs]: There is no requirement to delete customers

File: BangazonAPI/Controllers/CustomersController.cs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+                sql = $@"
+            SELECT
+                c.Id,
+                c.FirstName,
+                c.LastName
+            FROM Customer c
+              WHERE c.Id = {id}
+            ";
+            }
+            Console.WriteLine(sql);
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomersController.cs]: Remove `Console.WriteLine()`

File: BangazonAPI/Controllers/CustomersController.cs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+                if (q != null || _include != null)
+                {
+                    string isQ = $@"
+                    AND c.FirstName LIKE '%{q}%'
+                    OR c.LastName LIKE '%{q}%'
+                ";
+                    sql = $"{sql} {isQ}";
+                }
+            }
+            Console.WriteLine(sql);
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomersController.cs]: Remove `Console.WriteLine()`

File: BangazonAPI/Controllers/CustomersController.cs [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            {
+                return new SqlConnection(_config.GetConnectionString("DefaultConnection"));
+            }
+        }
+        // GET api/customers?q=Taco
+        [HttpGet]
+        public async Task<IActionResult> Get(string q, string _include)
+        {
+            string sql;
+            if (_include != null && _include.Contains("payments"))
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomersController.cs]: What happens if I enter a URL such as http://localhost:5000/api/customers?_include=gimme_the_payments ?

File: BangazonAPI/BangazonAPI.csproj [hunk 5]
```diff
@@ -19,6 +19,7 @@
     <PackageReference Include="Dapper" Version="1.50.5" />
     <PackageReference Include="Microsoft.AspNetCore.App" />
     <PackageReference Include="Microsoft.AspNetCore.Razor.Design" Version="2.1.2" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="2.1.1" />
```
Comment (askingalot) [inline BangazonAPI/BangazonAPI.csproj]: What is this?

File: BangazonAPI/Controllers/CustomersController.cs [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+                if (q != null)
+                {
+                    string isQ = $@"
+                    AND c.FirstName LIKE '%{q}%'
+                    OR c.LastName LIKE '%{q}%'
+                ";
+                    sql = $"{sql} {isQ}";
+                }
+            }
+            else if (_include != null && _include.Contains("products"))
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomersController.cs]: When `_include=products` is specified and a customer has not created any products, that customer is not being returned. I would expect to see the customer with an empty list of products.

**Review by askingalot — CHANGES_REQUESTED** [review]
In addition to resolving my other comments, please include the README that we discussed.

**Comment by jeremiah3643** [comment]
Changes that @askingalot requested have been made. PR has been updated.

**Review by askingalot — APPROVED** [review]

**Review by KHardt — APPROVED** [review]
Looks good. Tested GET, PUT, POST, q/search, include products and include payments.

**Pull request merged**